doc/bulk_counter/bulk_counter.md

```diff
@@ -0,0 +1,144 @@
+# SONiC Bulk Counter Design #
+
+## Table of Content
+
+### Revision
+
+ | Rev |     Date    |       Author       | Change Description                |
+ |:---:|:-----------:|:------------------:|-----------------------------------|
+ | 0.1 |             |      Junchao Chen  | Initial version                   |
+
+### Scope
+
+This document is the design document for bulk counter feature on SONiC.
+
+### Definitions/Abbreviations
+
+N/A
+
+### Overview
+
+PR https://github.com/opencomputeproject/SAI/pull/1352/files introduced new SAI APIs that supports bulk stats:
+
+- sai_bulk_object_get_stats
+- sai_bulk_object_clear_stats
+
+SONiC flex counter infrastructure shall utilize bulk stats API to gain better performance. This document discusses how to integrate these two new APIs to SONiC. 
+
+### Requirements
+
+- Syncd shall use bulk stats APIs based on object type. E.g. for a counter group that queries queue and pg stats, queue stats support bulk while pg stats does not, in that case queue stats shall use bulk API, pg stats shall use non bulk API
+- For a certain object type in a counter group, it shall use bulk stats only if:
+  - The stats capability for each counter IDs shall include SAI_STATS_MODE_BULK_READ, SAI_STATS_MODE_BULK_CLEAR, SAI_STATS_MODE_BULK_READ_AND_CLEAR (avoid handling stats mode change)
```

Comment:
Must include all three, or some (for example BULK_READ and BULK_CLEAR) ? Even if it only includes BULK_READ, can it do bulk get and do individual clear?

Reply:
I will fix: stats capability should match flex counter group mode.